File: python/cryptology/classical/substitution/monoalphabetic/keyword.py
"""
Keyword Cipher implementation.

The keyword cipher is a monoalphabetic substitution cipher where a keyword is used
to create the cipher alphabet. The keyword (with duplicates removed) is written first,
followed by the remaining letters of the alphabet in order.
"""

# Default English alphabet (lowercase only)
DEFAULT_ALPHABET = "abcdefghijklmnopqrstuvwxyz"


def _create_cipher_alphabet(keyword: str, alphabet: str) -> str:
    """
    Create cipher alphabet from keyword.
    
    Args:
        keyword: The keyword to use (will be converted to lowercase)
        alphabet: The base alphabet
    
    Returns:
        The cipher alphabet with keyword first, then remaining letters
    """
    keyword = keyword.lower()
    seen = set()
    cipher_alphabet = []
    
    # Add keyword letters (removing duplicates)
    for char in keyword:
        if char in alphabet and char not in seen:
            cipher_alphabet.append(char)
            seen.add(char)
    
    # Add remaining alphabet letters
    for char in alphabet:
        if char not in seen:
            cipher_alphabet.append(char)
    
    return ''.join(cipher_alphabet)


def encrypt(plaintext: str, keyword: str, alphabet: str = DEFAULT_ALPHABET) -> str:
    """
    Encrypt plaintext using Keyword cipher.
    
    Args:
        plaintext: The text to encrypt (will be converted to lowercase)
        keyword: The keyword to use for creating cipher alphabet
        alphabet: The alphabet to use for encryption (default: English lowercase)
    
    Returns:
        The encrypted ciphertext
    
    Example:
        >>> encrypt("HELLO", "zebra")
        'dclla'
    """
    # Convert input to lowercase
    plaintext = plaintext.lower()
    
    # Create cipher alphabet from keyword
    cipher_alphabet = _create_cipher_alphabet(keyword, alphabet)
    
    result = []
    for char in plaintext:
        if char in alphabet:
            # Find position in plain alphabet
            pos = alphabet.index(char)
            # Replace with character from cipher alphabet
            result.append(cipher_alphabet[pos])
        else:
            # Keep characters not in alphabet unchanged
            result.append(char)
    
    return ''.join(result)


def decrypt(ciphertext: str, keyword: str, alphabet: str = DEFAULT_ALPHABET) -> str:
    """
    Decrypt ciphertext using Keyword cipher.
    
    Args:
        ciphertext: The text to decrypt (will be converted to lowercase)
        keyword: The keyword that was used for encryption
        alphabet: The alphabet that was used for encryption (default: English lowercase)
    
    Returns:
        The decrypted plaintext
    
    Example:
        >>> decrypt("dclla", "zebra")
        'hello'
    """
    # Convert input to lowercase
    ciphertext = ciphertext.lower()
    
    # Create cipher alphabet from keyword
    cipher_alphabet = _create_cipher_alphabet(keyword, alphabet)
    
    result = []
    for char in ciphertext:
        if char in cipher_alphabet:
            # Find position in cipher alphabet
            pos = cipher_alphabet.index(char)
            # Replace with character from plain alphabet
            result.append(alphabet[pos])
        else:
            # Keep characters not in alphabet unchanged
            result.append(char)
    
    return ''.join(result)

File: c/src/classical/substitution/monoalphabetic/keyword.c
/**
 * @file keyword.c
 * @brief Keyword Cipher implementation
 */

#include "cryptology/classical/substitution/monoalphabetic/keyword.h"
#include <string.h>
#include <ctype.h>
#include <stdbool.h>

#define DEFAULT_ALPHABET "abcdefghijklmnopqrstuvwxyz"
#define MAX_ALPHABET_SIZE 256

static int find_char_in_alphabet(char c, const char *alphabet) {
    const char *pos = strchr(alphabet, c);
    return pos ? (int)(pos - alphabet) : -1;
}

static int create_cipher_alphabet(const char *keyword, const char *alphabet,
                                  char *cipher_alphabet, size_t cipher_size) {
    if (!keyword || !alphabet || !cipher_alphabet) {
        return -1;
    }
    
    size_t alphabet_len = strlen(alphabet);
    if (alphabet_len >= cipher_size) {
        return -1;
    }
    
    bool seen[MAX_ALPHABET_SIZE] = {false};
    size_t cipher_pos = 0;
    
    // Add keyword letters (removing duplicates)
    for (size_t i = 0; keyword[i] != '\0'; i++) {
        char c = tolower((unsigned char)keyword[i]);
        int pos = find_char_in_alphabet(c, alphabet);
        
        if (pos != -1 && !seen[pos]) {
            cipher_alphabet[cipher_pos++] = c;
            seen[pos] = true;
        }
    }
    
    // Add remaining alphabet letters
    for (size_t i = 0; i < alphabet_len; i++) {
        if (!seen[i]) {
            cipher_alphabet[cipher_pos++] = alphabet[i];
        }
    }
    
    cipher_alphabet[cipher_pos] = '\0';
    return 0;
}

int keyword_encrypt(const char *plaintext, const char *keyword, const char *alphabet,
                    char *result, size_t result_size) {
    if (!plaintext || !keyword || !result || result_size == 0) {
        return -1;
    }
    
    if (!alphabet) {
        alphabet = DEFAULT_ALPHABET;
    }
    
    char cipher_alphabet[MAX_ALPHABET_SIZE];
    if (create_cipher_alphabet(keyword, alphabet, cipher_alphabet, MAX_ALPHABET_SIZE) != 0) {
        return -1;
    }
    
    size_t plaintext_len = strlen(plaintext);
    if (plaintext_len >= result_size) {
        return -1;
    }
    
    for (size_t i = 0; i < plaintext_len; i++) {
        char c = tolower((unsigned char)plaintext[i]);
        int pos = find_char_in_alphabet(c, alphabet);
        
        if (pos != -1) {
            result[i] = cipher_alphabet[pos];
        } else {
            result[i] = plaintext[i];
        }
    }
    
    result[plaintext_len] = '\0';
    return 0;
}

int keyword_decrypt(const char *ciphertext, const char *keyword, const char *alphabet,
                    char *result, size_t result_size) {
    if (!ciphertext || !keyword || !result || result_size == 0) {
        return -1;
    }
    
    if (!alphabet) {
        alphabet = DEFAULT_ALPHABET;
    }
    
    char cipher_alphabet[MAX_ALPHABET_SIZE];
    if (create_cipher_alphabet(keyword, alphabet, cipher_alphabet, MAX_ALPHABET_SIZE) != 0) {
        return -1;
    }
    
    size_t ciphertext_len = strlen(ciphertext);
    if (ciphertext_len >= result_size) {
        return -1;
    }
    
    for (size_t i = 0; i < ciphertext_len; i++) {
        char c = tolower((unsigned char)ciphertext[i]);
        int pos = find_char_in_alphabet(c, cipher_alphabet);
        
        if (pos != -1) {
            result[i] = alphabet[pos];
        } else {
            result[i] = ciphertext[i];
        }
    }
    
    result[ciphertext_len] = '\0';
    return 0;
}

